File: essentia_core/bridge/parameter_map.cpp
#include "parameter_map.h"

ParameterMapBridge::ParameterMapBridge()
{
    _parameter_map = new essentia::ParameterMap();
}

ParameterMapBridge::~ParameterMapBridge()
{
    delete _parameter_map;
}

void ParameterMapBridge::add_string(rust::Str key, rust::Str value)
{
    std::string str_key(key);
    std::string str_value(value);
    _parameter_map->add(str_key, essentia::Parameter(str_value));
}

void ParameterMapBridge::add_real(rust::Str key, float value)
{
    std::string str_key(key);
    _parameter_map->add(str_key, essentia::Parameter(value));
}

void ParameterMapBridge::add_int(rust::Str key, int value)
{
    std::string str_key(key);
    _parameter_map->add(str_key, essentia::Parameter(value));
}

void ParameterMapBridge::add_bool(rust::Str key, bool value)
{
    std::string str_key(key);
    _parameter_map->add(str_key, essentia::Parameter(value));
}

const essentia::ParameterMap &ParameterMapBridge::get_parameter_map() const
{
    return *_parameter_map;
}

std::unique_ptr<ParameterMapBridge> create_parameter_map()
{
    return std::make_unique<ParameterMapBridge>();
}

File: essentia_core/src/parameter_map.rs
use std::pin::Pin;

use cxx::UniquePtr;

use crate::ffi::{self, ParameterMapBridge};

pub trait ParameterValue {
    fn set_parameter(&self, parameter_map: Pin<&mut ParameterMapBridge>, key: &str);
}

impl ParameterValue for &str {
    fn set_parameter(&self, parameter_map: Pin<&mut ParameterMapBridge>, key: &str) {
        parameter_map.add_string(key, self);
    }
}

impl ParameterValue for f32 {
    fn set_parameter(&self, parameter_map: Pin<&mut ParameterMapBridge>, key: &str) {
        parameter_map.add_real(key, *self);
    }
}

impl ParameterValue for i32 {
    fn set_parameter(&self, parameter_map: Pin<&mut ParameterMapBridge>, key: &str) {
        parameter_map.add_int(key, *self);
    }
}

impl ParameterValue for bool {
    fn set_parameter(&self, parameter_map: Pin<&mut ParameterMapBridge>, key: &str) {
        parameter_map.add_bool(key, *self);
    }
}

pub struct Parameters {
    pub(crate) inner: UniquePtr<ffi::ParameterMapBridge>,
}

impl Parameters {
    pub fn new() -> Self {
        let parameter_map = ffi::create_parameter_map();
        Self {
            inner: parameter_map,
        }
    }

    pub fn add_parameter(&mut self, key: &str, value: impl ParameterValue) {
        let parameter_map = self.inner.pin_mut();

        value.set_parameter(parameter_map, key);
    }
}
